src/cls/async-hooks.ts

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as asyncHooksModule from 'async_hooks';
+import {EventEmitter} from 'events';
+import * as shimmer from 'shimmer';
+
+import {CLS, Func} from './base';
+
+const EVENT_EMITTER_METHODS: Array<keyof EventEmitter> =
+    ['addListener', 'on', 'once', 'prependListener', 'prependOnceListener'];
+const WRAPPED = Symbol('context_wrapped');
+
+/**
+ * An implementation of continuation-local storage on top of the async_hooks
+ * module.
+ */
+export class AsyncHooksCLS<Context extends {}> implements CLS<Context> {
+  private current: {value: Context};
+  private contexts: {[id: number]: Context} = {};
+  private readonly defaultContext: Context;
+  private hook: asyncHooksModule.AsyncHook;
+  private enabled = false;
+
+  constructor(defaultContext: Context) {
+    this.defaultContext = defaultContext;
+    this.current = {value: this.defaultContext};
+    this.hook = (require('async_hooks') as typeof asyncHooksModule).createHook({
+      init: (id: number, type: string, triggerId: number, resource: {}) => {
+        this.contexts[id] = this.current.value;
+      },
+      before: (id: number) => {
+        if (this.contexts[id]) {
+          this.current.value = this.contexts[id];
+        }
+      },
+      destroy: (id: number) => {
+        delete this.contexts[id];
+      }
+    });
+  }
+
+  isEnabled(): boolean {
+    return this.enabled;
+  }
+
+  enable(): void {
+    this.current.value = this.defaultContext;
+    this.hook.enable();
+    this.enabled = true;
+  }
+
+  disable(): void {
+    this.current.value = this.defaultContext;
+    this.hook.disable();
+    this.enabled = false;
+  }
+
+  getContext(): Context {
+    return this.current.value;
+  }
+
+  setContext(value: Context): void {
+    this.current.value = value;
+  }
+
+  runWithNewContext<T>(fn: Func<T>): T {
+    const oldContext = this.current.value;
+    this.current.value = this.defaultContext;
+    const res = fn();
```

Comment:
Hmm.. I think there is a bug here if `fn` throws. This was present in the code even before your change. Let's fix this in a follow on. This should be:

```ts
try {
  return fn();
} finally {
  this.current.value = oldContext;
}
```

Can you add a test for this? (in the follow-on).

Reply:
Added this as a TODO.